OpenDreamRuntime/Procs/Native/DreamProcNativeRegex.cs

```diff
@@ -62,22 +63,70 @@ static class DreamProcNativeRegex {
             int start = arguments.GetArgument(2, "Start").GetValueAsInteger();
             int end = arguments.GetArgument(3, "End").GetValueAsInteger();
 
-            string haystackString = haystack.GetValueAsString();
+            if (!haystack.TryGetValueAsString(out var haystackString))
+            {
+                if (haystack.Value is null)
+                {
+                    return DreamValue.Null;
+                }
+                //TODO Check what actually happens
+                throw new ArgumentException("Bad regex haystack");
+            }
             string haystackSubstring = haystackString;
             if (end != 0) haystackSubstring = haystackString.Substring(0, end - start);
 
             if (replace.TryGetValueAsProc(out DreamProc replaceProc)) {
-                throw new NotImplementedException("Proc regex replacements are not implemented");
-            } else if (replace.TryGetValueAsString(out string replaceString))
+                return DoProcReplace(replaceProc);
+            }
+            if (replace.TryGetValueAsString(out string replaceString))
+            {
+                return DoTextReplace(replaceString);
+            }
+
+            if (replace.TryGetValueAsPath(out var procPath) && procPath.LastElement is not null)
+            {
+                var dreamMan = IoCManager.Resolve<IDreamManager>();
+                if (dreamMan.GlobalProcs.ContainsKey(procPath.LastElement))
+                {
+                    var proc = dreamMan.GlobalProcs[procPath.LastElement];
+                    return DoProcReplace(proc);
+
+                }
+            }
+
+            throw new ArgumentException("Replacement argument must be a string or a proc");
+
+            DreamValue DoProcReplace(DreamProc proc)
+            {
+                var match = dreamRegex.Regex.Match(haystackSubstring);
+                var captures = match.Captures;
+                List<DreamValue> args = new List<DreamValue>(captures.Count + 1);
+                args.Add(new DreamValue(match.Value));
+                foreach (var capture in captures)
+                {
+                    args.Add(capture is null ? DreamValue.Null : new DreamValue(capture));
+                }
```

Comment:
`captures` is a collection of `Capture` objects, not strings; you need to use their `Value` property. Even then you won't get the correct args because it won't give you all the groups captured by the regex.

`Groups` is the correct thing to use here, the first group will be the entire match and the rest will be what was matched by the capture groups (`( ... )` in the regex).
```suggestion
                var groups = match.Groups;
                List<DreamValue> args = new List<DreamValue>(groups.Count);
                foreach (Group group in groups)
                {
                    args.Add(new DreamValue(group.Value));
                }
```